runtime/storage.go

```diff
@@ -494,70 +366,63 @@ func (s *Storage) commit(inter *interpreter.Interpreter, commitContractUpdates b
 		s.commitContractUpdates(inter)
 	}
 
-	err := s.commitNewStorageMaps()
+	err := s.AccountStorageV1.commit()
 	if err != nil {
 		return err
 	}
 
-	// Migrate accounts that have write ops before calling PersistentSlabStorage.FastCommit().
-	err = s.migrateAccountsIfNeeded(inter)
+	err = s.AccountStorageV2.commit()
 	if err != nil {
 		return err
 	}
 
+	if s.Config.StorageFormatV2Enabled {
+		err = s.migrateV1AccountsToV2(inter)
+		if err != nil {
+			return err
+		}
+	}
+
 	// Commit the underlying slab storage's writes
 
-	size := s.PersistentSlabStorage.DeltasSizeWithoutTempAddresses()
+	slabStorage := s.PersistentSlabStorage
+
+	size := slabStorage.DeltasSizeWithoutTempAddresses()
 	if size > 0 {
 		inter.ReportComputation(common.ComputationKindEncodeValue, uint(size))
 		usage := common.NewBytesMemoryUsage(int(size))
-		common.UseMemory(s.memoryGauge, usage)
+		common.UseMemory(inter, usage)
 	}
 
-	deltas := s.PersistentSlabStorage.DeltasWithoutTempAddresses()
-	common.UseMemory(s.memoryGauge, common.NewAtreeEncodedSlabMemoryUsage(deltas))
+	deltas := slabStorage.DeltasWithoutTempAddresses()
+	common.UseMemory(inter, common.NewAtreeEncodedSlabMemoryUsage(deltas))
 
 	// TODO: report encoding metric for all encoded slabs
 	if deterministic {
-		return s.PersistentSlabStorage.FastCommit(runtime.NumCPU())
+		return slabStorage.FastCommit(runtime.NumCPU())
 	} else {
-		return s.PersistentSlabStorage.NondeterministicFastCommit(runtime.NumCPU())
+		return slabStorage.NondeterministicFastCommit(runtime.NumCPU())
 	}
 }
 
-func (s *Storage) commitNewStorageMaps() error {
-	if s.NewAccountStorageMapSlabIndices == nil {
-		return nil
-	}
+func (s *Storage) ScheduleV2Migration(address common.Address) {
+	s.scheduledV2Migrations = append(s.scheduledV2Migrations, address)
+}
 
-	for pair := s.NewAccountStorageMapSlabIndices.Oldest(); pair != nil; pair = pair.Next() {
-		var err error
-		errors.WrapPanic(func() {
-			err = s.Ledger.SetValue(
-				pair.Key.Address[:],
-				[]byte(pair.Key.Key),
-				pair.Value[:],
-			)
-		})
-		if err != nil {
-			return interpreter.WrappedExternalError(err)
+func (s *Storage) ScheduleV2MigrationForModifiedAccounts() {
```

Comment:
We now expose this (and the simple function `ScheduleV2Migration` above) to allow e.g. FVM to migrate accounts. Just like in the current feature branch, the migration is performed after committing program execution changes, and before serializing all changes.

We might not end up using this function, we'll discuss it in the meeting with the Execution team.